File: src/main/java/org/algohub/engine/judge/JudgeTask.java
package org.algohub.engine.judge;

import org.algohub.engine.type.InternalTestCase;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

class JudgeTask implements Callable<List<Boolean>> {

    private final Object clazz;
    private final Method method;
    private final InternalTestCase[] testCases;

    JudgeTask(Object clazz, Method method, InternalTestCase[] testCases) {
        this.clazz = clazz;
        this.method = method;
        this.testCases = testCases;
    }

    @Override
    public List<Boolean> call() throws Exception {
        return run();
    }

    List<Boolean> run() {
        ArrayList<Boolean> results = new ArrayList<>();

        for (final InternalTestCase internalTestCase : testCases) {
            final boolean isCorrect = judge(clazz, method, internalTestCase);

            results.add(isCorrect);
        }

        return results;
    }

    private static boolean judge(final Object clazz, final Method method,
                                 final InternalTestCase testCase) {
        final Object output;
        try {
            output = method.invoke(clazz, testCase.getInput());
        } catch (IllegalAccessException | InvocationTargetException e) {
            throw new IllegalStateException(e.getMessage());
        }

        return BetterObjects.equal(testCase.getOutput(), output);
    }
}

File: src/main/kotlin/com/jalgoarena/judge/JudgeTask.kt
package com.jalgoarena.judge

import com.jalgoarena.utils.BetterObjects
import com.jalgoarena.utils.ThrowableEnhancements
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.concurrent.Callable

internal class JudgeTask(private val clazz: Any, private val method: Method, private val testCases: Array<InternalTestCase>)
    : Callable<List<Boolean>>, ThrowableEnhancements {

    private val LOG = LoggerFactory.getLogger(this.javaClass)

    override fun call(): List<Boolean> = testCases.map { judge(clazz, method, it) }

    private fun judge(clazz: Any, method: Method, testCase: InternalTestCase): Boolean {

        val output: Any?
        val input: Array<Any?>

        try {
            input = testCase.input
            output = method.invoke(clazz, *input)
        } catch (e: IllegalAccessException) {
            val cause = getCause(e)
            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        } catch (e: InvocationTargetException) {
            val cause = getCause(e)

            if (cause is NullPointerException) {
                return false
            }

            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        } catch (e: IllegalArgumentException) {
            val cause = getCause(e)
            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        }

        return BetterObjects.equalForObjectsOrArrays(
                testCase.output,
                if (testCase.returnsVoid) input[0] else output
        )
    }
}
